edit_file tool

robotdad opened the pull request:
Suggested implementation for an edit_file per #183 

## Description
Adds a edit_file tool to filesystem.

## Server Details
<!-- If modifying an existing server or adding a new one, provide details -->
- Server: filesystem
- Changes to: tools

## Motivation and Context
Described in the issue, but write_file really isn't trustworthy for existing files.

## How Has This Been Tested?
Claude Desktop, not a ton of interactions with it yet. I tried it on some code in a git repo on Windows which surfaced some line ending issues it can handle now. 

## Breaking Changes
No.

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] New MCP Server
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [ ] I have read the [MCP Protocol Documentation](https://modelcontextprotocol.io)
- [x] My server follows MCP security best practices
- [x] I have updated the server's README accordingly
- [x] I have tested this with an LLM client
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [x] I have documented all environment variables and configuration options

## Additional context
<!-- Add any other context, implementation notes, or design decisions -->

jspahrsummers commented on src/filesystem/README.md:
```diff
@@ -36,6 +36,35 @@ Node.js server implementing Model Context Protocol (MCP) for filesystem operatio
+  - Inputs:
+    - `path` (string): File to edit
+    - `edits` (array): List of edit operations
+      - `startLine?` (number): Line number for edit (optional)
+      - `findAnchor?` (string): Text to locate edit position (optional)
+      - `anchorOffset` (number): Lines to offset from anchor (default: 0)
+      - `oldText` (string): Content to replace/verify
+      - `newText` (string): New content to insert
+      - `insertMode` (string): 'replace', 'before', or 'after' (default: 'replace')
+      - `beforeContext?` (string): Expected content before edit point (optional)
+      - `afterContext?` (string): Expected content after edit point (optional)
+      - `contextRadius` (number): Lines to check for context (default: 3)
+      - `verifyState` (boolean): Verify content matches before editing (default: true)
+      - `readBeforeEdit` (boolean): Refresh file state between edits (default: false)
+      - `dryRun` (boolean): Preview changes without applying them (default: false)
```
Most of these properties don't actually seem to exist

robotdad replied:
Apologies, the readme fell behind my commits when Claude shut me down last night. Fixed now.

jspahrsummers commented on src/filesystem/index.ts:
```diff
@@ -202,6 +216,61 @@ async function searchFiles(
+    // Calculate line number for reporting
+    const lineNumber = content.slice(0, pos).split(/\r?\n/).length;
+    
+    if (edit.dryRun) {
+      // Create git-style diff preview
+      const preview = [
+        `@@ line ${lineNumber} @@`,
+        '<<<<<<< ORIGINAL',
+        edit.oldText,
+        '=======',
+        edit.newText,
+        '>>>>>>> MODIFIED'
+      ].join('\n');
+      
+      previews.push({
+        original: edit.oldText,
+        modified: edit.newText,
+        lineNumber,
+        preview
+      });
+      continue;
+    }
+    
+    // Apply the edit
+    content = content.slice(0, pos) + edit.newText + content.slice(pos + edit.oldText.length);
+  }
+  
+  if (edits.some(e => e.dryRun)) {
+    return previews;
+  }
```
Since `dryRun` seems changes the logic _and_ the return type, can we create two different functions instead? We can reuse common bits between them instead of `if`/`else`-ing most of the logic in one function.

robotdad replied:
Looks like this is still a good suggestion after the refactoring I did last night. I can take a look at this later as I'll need to rerun all of my tests after making this change.

jspahrsummers commented on src/filesystem/index.ts:
```diff
@@ -106,6 +106,20 @@ const WriteFileArgsSchema = z.object({
+
+const EditFileArgsSchema = z.object({
+  path: z.string(),
+  edits: z.array(EditOperation),
```
IMO `dryRun` makes more sense as an argument here, rather than per-operation.

robotdad replied:
I'll take a look at this while working on the other dry run suggestion, thanks.

jspahrsummers reviewed (changes requested):
Thanks for the PR! I like this idea, just left some comments with suggestions.

robotdad commented:
@jspahrsummers thank you for the review and encouragement. I made some significant changes to the PR based on some testing last night. I have changed it to using a git style diff and substring matching informed by how Aider handles things. This made things a lot simpler and works a lot better with less special handling.  I used Claude to generate a bunch of test documents of likely file types this would be used with described in [TEST_DOCUMENTS.md](https://github.com/user-attachments/files/17996381/TEST_DOCUMENTS.md), you can regenerate them using that document and the file system tool. [TEST_RESULTS.md](https://github.com/user-attachments/files/17996380/TEST_RESULTS.md) records findings and it worked really well, the edit_file tool was used for editing that document while testing other than the initial creation. I've recorded the limitations there and added them to the read me, but I don't think they are significant concerns for the reasons stated. (boils down to there may be issues with mixed tabs/spaces)

sockdrawermoney commented:
this is awesome @robotdad, excited for this to land

robotdad commented:
I'm working through some things I've found issues with multiline edits. I'm eager to get this over the finish line... hopefully Claude doesn't keep cutting me off. 🤣

robotdad commented:
I moved to the diff package and have taken a lot of tips from how aider does things. This appears to be working a lot better. However, I'm still running into issues with real world code.  It's a lot closer though. Current changes are all pushed. Done for the night because Claude shut me down... I _pay_ for this kind of treatment? 🤣

robotdad commented:
I didn't go to bed. This looks like it is working pretty well now. 
[TEST_DOCUMENTS.md](https://github.com/user-attachments/files/18019953/TEST_DOCUMENTS.md), added some more complex cases for ts and python.
[TEST_RESULTS.md](https://github.com/user-attachments/files/18019961/TEST_RESULTS.md)

For transparency, though these all say passed it did have problems with some of the edits but noticed them in dry run and corrected itself. I'd like to figure out how to export that chat transcript and share it, but I'll look at that tomorrow. 

@jspahrsummers your comments are addressed in that that code isn't here anymore after better matching the aider approach to diffs.

jspahrsummers reviewed (approved):
Great, thanks for the updates!
